File: sdks/dotnet/AeonProcessorSdk/Runner.cs
// Aeon Processor SDK — WebSocket T4 runner with AWPP handshake.

using System.Net.WebSockets;
using System.Text;
using System.Text.Json;

namespace Aeon.ProcessorSdk;

/// <summary>Configuration for the WebSocket processor runner.</summary>
public sealed class RunnerConfig
{
    /// <summary>Processor name (registered with the engine).</summary>
    public string Name { get; set; } = "dotnet-processor";

    /// <summary>Processor version string.</summary>
    public string Version { get; set; } = "1.0.0";

    /// <summary>Path to 32-byte ED25519 seed file. Mutually exclusive with PrivateKey.</summary>
    public string? PrivateKeyPath { get; set; }

    /// <summary>32-byte ED25519 seed. Mutually exclusive with PrivateKeyPath.</summary>
    public byte[]? PrivateKey { get; set; }

    /// <summary>Requested pipeline names.</summary>
    public List<string> Pipelines { get; set; } = new();

    /// <summary>Transport codec: "msgpack" (default) or "json".</summary>
    public string CodecName { get; set; } = "msgpack";

    /// <summary>Processor registration (per-event or batch).</summary>
    public ProcessorRegistration Processor { get; set; } = ProcessorRegistration.PerEvent(_ => new List<Output>());

    /// <summary>Maximum batch size to request from the engine.</summary>
    public int MaxBatchSize { get; set; } = 1000;

    /// <summary>Binding mode: "dedicated" or "shared".</summary>
    public string Binding { get; set; } = "dedicated";
}

/// <summary>
/// Runs a processor connected to the Aeon engine via WebSocket (T4 transport).
/// Implements the full AWPP handshake: Challenge → Register → Accepted,
/// then enters the heartbeat + batch processing loop.
/// </summary>
public static class Runner
{
    /// <summary>
    /// Connect to the Aeon engine and run the processor loop until cancelled.
    /// </summary>
    /// <param name="url">WebSocket URL, e.g. ws://localhost:4471/api/v1/processors/connect</param>
    /// <param name="config">Runner configuration.</param>
    /// <param name="ct">Cancellation token for graceful shutdown.</param>
    public static async Task RunAsync(string url, RunnerConfig config, CancellationToken ct = default)
    {
        // Create signer
        using var signer = config.PrivateKeyPath != null
            ? Signer.FromFile(config.PrivateKeyPath)
            : config.PrivateKey != null
                ? Signer.FromSeed(config.PrivateKey)
                : Signer.Generate();

        var codec = new Codec(config.CodecName);

        using var ws = new ClientWebSocket();
        await ws.ConnectAsync(new Uri(url), ct);

        var recvBuf = new byte[4 * 1024 * 1024]; // 4MB receive buffer
        bool handshakeComplete = false;
        bool batchSigningEnabled = false;
        int heartbeatIntervalMs = 10000;

        // Heartbeat timer (started after handshake)
        Timer? heartbeatTimer = null;

        try
        {
            while (ws.State == WebSocketState.Open && !ct.IsCancellationRequested)
            {
                var result = await ws.ReceiveAsync(recvBuf, ct);

                if (result.MessageType == WebSocketMessageType.Close)
                    break;

                var msgData = new ReadOnlyMemory<byte>(recvBuf, 0, result.Count);

                if (!handshakeComplete || result.MessageType == WebSocketMessageType.Text)
                {
                    // Text frame — AWPP control message
                    var text = Encoding.UTF8.GetString(msgData.Span);
                    var msg = JsonDocument.Parse(text);
                    var msgType = msg.RootElement.GetProperty("type").GetString();

                    if (msgType == "challenge")
                    {
                        var nonce = msg.RootElement.GetProperty("nonce").GetString()!;
                        var reg = new Dictionary<string, object?>
                        {
                            ["type"] = "register",
                            ["protocol"] = "awpp/1",
                            ["transport"] = "websocket",
                            ["name"] = config.Name,
                            ["version"] = config.Version,
                            ["public_key"] = signer.PublicKeyFormatted,
                            ["challenge_signature"] = signer.SignChallenge(nonce),
                            ["oauth_token"] = null,
                            ["capabilities"] = new[] { "batch" },
                            ["max_batch_size"] = config.MaxBatchSize,
                            ["transport_codec"] = config.CodecName,
                            ["requested_pipelines"] = config.Pipelines.ToArray(),
                            ["binding"] = config.Binding,
                        };
                        var regJson = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(reg));
                        await ws.SendAsync(regJson, WebSocketMessageType.Text, true, ct);
                    }
                    else if (msgType == "accepted")
                    {
                        handshakeComplete = true;
                        if (msg.RootElement.TryGetProperty("heartbeat_interval_ms", out var hb))
                            heartbeatIntervalMs = hb.GetInt32();
                        if (msg.RootElement.TryGetProperty("batch_signing", out var bs))
                            batchSigningEnabled = bs.GetBoolean();

                        // Start heartbeat
                        heartbeatTimer = new Timer(async _ =>
                        {
                            if (ws.State != WebSocketState.Open) return;
                            try
                            {
                                var hbMsg = JsonSerializer.Serialize(new Dictionary<string, object>
                                {
                                    ["type"] = "heartbeat",
                                    ["timestamp_ms"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                                });
                                var hbBytes = Encoding.UTF8.GetBytes(hbMsg);
                                await ws.SendAsync(hbBytes, WebSocketMessageType.Text, true, CancellationToken.None);
                            }
                            catch { /* heartbeat failure is non-fatal */ }
                        }, null, heartbeatIntervalMs, heartbeatIntervalMs);
                    }
                    else if (msgType == "rejected")
                    {
                        var code = msg.RootElement.TryGetProperty("code", out var c) ? c.GetString() : "unknown";
                        var message = msg.RootElement.TryGetProperty("message", out var m) ? m.GetString() : "";
                        throw new InvalidOperationException($"AWPP rejected: [{code}] {message}");
                    }
                    else if (msgType == "drain")
                    {
                        break;
                    }
                    continue;
                }

                // Binary frame — batch data with routing header
                var frame = DataFrameWire.Parse(msgData.Span);
                var batch = BatchWire.DecodeBatchRequest(frame.BatchData, codec);

                // Process events
                List<List<Output>> outputsPerEvent;
                if (config.Processor.BatchProcessFn != null)
                {
                    outputsPerEvent = config.Processor.BatchProcessFn(batch.Events);
                }
                else if (config.Processor.ProcessFn != null)
                {
                    outputsPerEvent = batch.Events
                        .Select(e => config.Processor.ProcessFn(e))
                        .ToList();
                }
                else
                {
                    outputsPerEvent = batch.Events.Select(_ => new List<Output>()).ToList();
                }

                // Encode and send response
                var responseBatch = BatchWire.EncodeBatchResponse(
                    batch.BatchId, outputsPerEvent, codec, signer, batchSigningEnabled);
                var responseFrame = DataFrameWire.Build(
                    frame.PipelineName, frame.Partition, responseBatch);
                await ws.SendAsync(responseFrame, WebSocketMessageType.Binary, true, ct);
            }
        }
        finally
        {
            heartbeatTimer?.Dispose();
            if (ws.State == WebSocketState.Open)
            {
                try { await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "shutdown", CancellationToken.None); }
                catch { /* best-effort close */ }
            }
        }
    }
}

File: sdks/java/src/main/java/io/aeon/processor/Runner.java
package io.aeon.processor;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

/**
 * AWPP WebSocket runner (T4 transport).
 * Handles the full AWPP handshake, heartbeat loop, and binary frame processing.
 */
public final class Runner {

    private Runner() {}

    /**
     * Runner configuration.
     */
    public record RunnerConfig(
        String name,
        String version,
        byte[] privateKey,
        List<String> pipelines,
        String codec,
        Processor processor,
        BatchProcessor batchProcessor
    ) {
        public RunnerConfig {
            if (name == null || name.isEmpty()) throw new IllegalArgumentException("name required");
            if (version == null || version.isEmpty()) throw new IllegalArgumentException("version required");
            if (pipelines == null || pipelines.isEmpty()) throw new IllegalArgumentException("pipelines required");
            if (processor == null && batchProcessor == null)
                throw new IllegalArgumentException("processor or batchProcessor required");
            if (codec == null || codec.isEmpty()) codec = "json";
        }
    }

    /**
     * Connect to an Aeon engine via WebSocket and run the AWPP protocol.
     * Blocks until the connection is closed or an error occurs.
     */
    public static void run(String url, RunnerConfig config) {
        var signer = config.privateKey() != null
                ? Signer.fromSeed(config.privateKey())
                : Signer.generate();
        var codecInst = new Codec();
        var closeLatch = new CountDownLatch(1);
        var batchSigning = new AtomicBoolean(false);
        var heartbeatInterval = new AtomicReference<Long>(10000L);
        var sessionEstablished = new AtomicBoolean(false);
        var wsRef = new AtomicReference<WebSocket>();
        var textBuffer = new StringBuilder();

        var listener = new WebSocket.Listener() {

            @Override
            public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
                textBuffer.append(data);
                if (!last) {
                    webSocket.request(1);
                    return CompletableFuture.completedFuture(null);
                }

                String message = textBuffer.toString();
                textBuffer.setLength(0);
                webSocket.request(1);

                try {
                    var json = Codec.parseObject(message);
                    String type = (String) json.get("type");

                    switch (type) {
                        case "challenge" -> handleChallenge(webSocket, json, config, signer);
                        case "accepted" -> {
                            String sessionId = (String) json.get("session_id");
                            Object hbInterval = json.get("heartbeat_interval_ms");
                            Object bs = json.get("batch_signing");
                            if (hbInterval != null) {
                                heartbeatInterval.set(((Number) hbInterval).longValue());
                            }
                            if (bs != null && (bs.equals(Boolean.TRUE) || "true".equals(bs.toString()))) {
                                batchSigning.set(true);
                            }
                            sessionEstablished.set(true);
                            System.out.println("[AEON] Session established: " + sessionId);
                            // Start heartbeat thread
                            startHeartbeat(webSocket, heartbeatInterval.get(), closeLatch);
                        }
                        case "heartbeat" -> { /* Server heartbeat — acknowledged implicitly */ }
                        case "error" -> {
                            System.err.println("[AEON] Error from server: " + json.get("message"));
                            closeLatch.countDown();
                        }
                        default -> System.err.println("[AEON] Unknown message type: " + type);
                    }
                } catch (Exception e) {
                    System.err.println("[AEON] Error handling text message: " + e.getMessage());
                }

                return CompletableFuture.completedFuture(null);
            }

            @Override
            public CompletionStage<?> onBinary(WebSocket webSocket, ByteBuffer data, boolean last) {
                webSocket.request(1);

                try {
                    byte[] frameBytes = new byte[data.remaining()];
                    data.get(frameBytes);

                    // Parse data frame (routing header)
                    var frame = Wire.parseDataFrame(frameBytes);

                    // Decode batch request
                    var batchReq = Wire.decodeBatchRequest(frame.batchData(), codecInst);

                    // Process events
                    List<List<Output>> outputsPerEvent;
                    if (config.batchProcessor() != null) {
                        outputsPerEvent = config.batchProcessor().processBatch(batchReq.events());
                    } else {
                        outputsPerEvent = new ArrayList<>();
                        for (var event : batchReq.events()) {
                            outputsPerEvent.add(config.processor().process(event));
                        }
                    }

                    // Encode batch response
                    byte[] responseData = Wire.encodeBatchResponse(
                            batchReq.batchId(),
                            outputsPerEvent,
                            codecInst,
                            signer,
                            batchSigning.get()
                    );

                    // Wrap in data frame
                    byte[] responseFrame = Wire.buildDataFrame(
                            frame.pipelineName(),
                            frame.partition(),
                            responseData
                    );

                    // Send binary response
                    webSocket.sendBinary(ByteBuffer.wrap(responseFrame), true);

                } catch (Exception e) {
                    System.err.println("[AEON] Error processing binary frame: " + e.getMessage());
                    e.printStackTrace();
                }

                return CompletableFuture.completedFuture(null);
            }

            @Override
            public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
                System.out.println("[AEON] Connection closed: " + statusCode + " " + reason);
                closeLatch.countDown();
                return CompletableFuture.completedFuture(null);
            }

            @Override
            public void onError(WebSocket webSocket, Throwable error) {
                System.err.println("[AEON] WebSocket error: " + error.getMessage());
                closeLatch.countDown();
            }
        };

        try {
            var client = HttpClient.newHttpClient();
            var ws = client.newWebSocketBuilder()
                    .subprotocols("awpp")
                    .buildAsync(URI.create(url), listener)
                    .join();
            wsRef.set(ws);

            System.out.println("[AEON] Connected to " + url);
            ws.request(1);

            // Block until connection closes
            closeLatch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            System.err.println("[AEON] Interrupted");
        } catch (Exception e) {
            System.err.println("[AEON] Connection failed: " + e.getMessage());
        }
    }

    private static void handleChallenge(
            WebSocket ws,
            Map<String, Object> challenge,
            RunnerConfig config,
            Signer signer
    ) {
        String nonce = (String) challenge.get("nonce");
        String signature = signer.signChallenge(nonce);

        // Build register message
        var sb = new StringBuilder(512);
        sb.append("{\"type\":\"register\"");
        sb.append(",\"protocol\":\"awpp/1\"");
        sb.append(",\"transport\":\"websocket\"");
        sb.append(",\"name\":\"").append(jsonEscape(config.name())).append('"');
        sb.append(",\"version\":\"").append(jsonEscape(config.version())).append('"');
        sb.append(",\"public_key\":\"").append(jsonEscape(signer.publicKeyFormatted())).append('"');
        sb.append(",\"challenge_signature\":\"").append(signature).append('"');
        sb.append(",\"capabilities\":[\"batch\"]");
        sb.append(",\"max_batch_size\":1000");
        sb.append(",\"transport_codec\":\"").append(jsonEscape(config.codec())).append('"');
        sb.append(",\"requested_pipelines\":[");
        for (int i = 0; i < config.pipelines().size(); i++) {
            if (i > 0) sb.append(',');
            sb.append('"').append(jsonEscape(config.pipelines().get(i))).append('"');
        }
        sb.append(']');
        sb.append(",\"binding\":\"dedicated\"");
        sb.append('}');

        ws.sendText(sb.toString(), true);
    }

    private static void startHeartbeat(WebSocket ws, long intervalMs, CountDownLatch closeLatch) {
        var thread = new Thread(() -> {
            try {
                while (closeLatch.getCount() > 0) {
                    Thread.sleep(intervalMs);
                    if (closeLatch.getCount() == 0) break;
                    long ts = System.currentTimeMillis();
                    String hb = "{\"type\":\"heartbeat\",\"timestamp_ms\":" + ts + "}";
                    ws.sendText(hb, true);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } catch (Exception e) {
                // Connection likely closed
            }
        }, "aeon-heartbeat");
        thread.setDaemon(true);
        thread.start();
    }

    private static String jsonEscape(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}
